File: legado-php/scripts/autoteste.php
<?php
/**
 * autoteste.php — os validadores realmente pegam o que prometem?
 *
 * Uso:  php scripts/autoteste.php
 * Saída: 0 se todos os testes passam, 1 se algum falha.
 *
 * POR QUE ISTO EXISTE
 * Verificador que não reprova nada é indistinguível de verificador quebrado — os dois
 * imprimem "tudo certo". E verificador quebra em silêncio: basta um regex que deixou de
 * casar, uma pasta renomeada, um caminho com separador diferente.
 *
 * Num caso real, um contador passou meses devolvendo 73 em vez de 71 porque comparava
 * caminho com `/` numa máquina que devolvia `\`. O script terminava com sucesso. Só foi
 * percebido porque o número certo estava escrito num documento.
 *
 * Este script quebra o projeto DE PROPÓSITO, uma coisa por vez, e confere que o
 * validador certo reclama. É o teste do alarme de incêndio: você aperta o botão.
 *
 * COMO FUNCIONA — e por que é seguro
 * Nada é feito no seu projeto. Tudo acontece numa CÓPIA em pasta temporária, que é
 * apagada no fim. Se o script for interrompido, a cópia fica lá e não afeta nada.
 *
 * Rode depois de:
 *   - instalar o padrão      (confirma que a instalação ficou funcional)
 *   - mexer no padrao.json   (contador novo pega mesmo o que você quer?)
 *   - atualizar o kit
 */

require_once __DIR__ . '/lib-padrao.php';

$origem = raiz();
$tmp    = sys_get_temp_dir() . '/autoteste-padrao-' . getmypid();

/** Cópia recursiva, pulando o que não importa e o que é pesado. */
function copiar(string $de, string $para): void
{
    @mkdir($para, 0775, true);
    foreach (scandir($de) ?: [] as $item) {
        if ($item === '.' || $item === '..' || $item === '.git'
            || $item === 'node_modules' || $item === 'vendor') {
            continue;
        }
        $o = $de . '/' . $item;
        $d = $para . '/' . $item;
        is_dir($o) ? copiar($o, $d) : @copy($o, $d);
    }
}

function apagar(string $caminho): void
{
    if (!is_dir($caminho)) {
        @unlink($caminho);
        return;
    }
    foreach (scandir($caminho) ?: [] as $i) {
        if ($i !== '.' && $i !== '..') {
            apagar($caminho . '/' . $i);
        }
    }
    @rmdir($caminho);
}

/**
 * Escreve um arquivo criando o diretório se preciso.
 *
 * Existe por causa de um bug deste próprio script: ele escrevia direto em
 * `docs/arquitetura/`, que no projeto de origem existia mas num CLONE não — git não
 * versiona diretório vazio. Quatro testes falhavam no CI e passavam na máquina de quem
 * escreveu, com aviso do PHP no meio da saída.
 *
 * É a mesma armadilha que o gerador já tinha caído. Vale para qualquer coisa que
 * escreva em pasta do repositório: **nunca presuma que o diretório existe do outro lado.**
 */
function escrever(string $arquivo, string $conteudo): void
{
    @mkdir(dirname($arquivo), 0775, true);
    file_put_contents($arquivo, $conteudo);
}

/** Frontmatter mínimo e válido, para os testes de cadeia montarem pai e filho. */
function cabecalhoDeTeste(string $id, string $tipo, string $status, array $pais): string
{
    $fm  = "---\nid: {$id}\ntipo: {$tipo}\nprojeto: " . config()['projeto']
         . "\ntitulo: documento de teste do autoteste\nstatus: {$status}\n";
    $fm .= "verificado_em: 2026-01-01\nverificado_commit: 0000000\n";
    if ($pais) {
        $fm .= "deriva_de:\n";
        foreach ($pais as $p) {
            $fm .= "  - {$p}\n";
        }
    }
    return $fm . "---\n\n# documento de teste\n";
}

/** Roda um validador na cópia e devolve [código de saída, saída]. */
function rodar(string $tmp, string $script): array
{
    /*
     * Os scripts rodam DO PACOTE, com o diretório corrente na cópia. Antes o comando
     * era `php $tmp/scripts/<script>` — funcionava quando o kit era copiado para o
     * projeto, e quebraria via Composer, onde a cópia do projeto não tem scripts/
     * (eles moram no vendor, que a cópia pula de propósito).
     */
    $saida = [];
    $rc = 0;
    exec('cd ' . escapeshellarg($tmp) . ' && php ' . escapeshellarg(__DIR__ . '/' . $script) . ' 2>&1', $saida, $rc);
    return [$rc, implode("\n", $saida)];
}

// ══════════════════════════════════════════════════════════════ os testes
/*
 * Cada teste: um nome, uma função que estraga a cópia, o validador que DEVE reclamar,
 * e um trecho que precisa aparecer na saída.
 *
 * `esperado` = 1 significa "tem que reprovar". `esperado` = 0 significa "NÃO pode
 * reprovar" — esses são tão importantes quanto os outros, porque validador que reprova
 * demais é abandonado tão rápido quanto validador que não reprova nada.
 */
$testes = [
    [
        'nome'     => 'âncora apontando para arquivo inexistente',
        'quebra'   => fn (string $t) => escrever($t . '/docs/arquitetura/_teste.md',
            "---\nid: TESTE-ANCORA\ntipo: hld\nprojeto: " . config()['projeto']
            . "\ntitulo: t\nstatus: rascunho\nancoras:\n  - caminho/que/nao/existe.php\n---\n# t\n"),
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'âncora quebrada',
    ],
    [
        'nome'     => 'tipo fora do vocabulário',
        'quebra'   => fn (string $t) => escrever($t . '/docs/arquitetura/_teste.md',
            "---\nid: TESTE-VOCAB\ntipo: inventado\nprojeto: " . config()['projeto']
            . "\ntitulo: t\nstatus: rascunho\n---\n# t\n"),
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'fora do vocabulário',
    ],
    [
        'nome'     => 'nome de projeto divergente do padrao.json',
        'quebra'   => fn (string $t) => escrever($t . '/docs/arquitetura/_teste.md',
            "---\nid: TESTE-PROJ\ntipo: hld\nprojeto: outro-repositorio\n"
            . "titulo: t\nstatus: rascunho\n---\n# t\n"),
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'não bate com o',
    ],
    [
        'nome'     => 'id duplicado',
        'quebra'   => function (string $t) {
            $fm = "---\nid: TESTE-DUP\ntipo: hld\nprojeto: " . config()['projeto']
                . "\ntitulo: t\nstatus: rascunho\n---\n# t\n";
            escrever($t . '/docs/arquitetura/_a.md', $fm);
            escrever($t . '/docs/arquitetura/_b.md', $fm);
        },
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'duplicado',
    ],
    [
        'nome'     => 'arquivo derivado editado à mão',
        'quebra'   => function (string $t) {
            $dir = $t . '/' . trim(config()['gerado']['diretorio'], '/');
            foreach (glob($dir . '/*.md') ?: [] as $g) {
                file_put_contents($g, file_get_contents($g) . "\nlinha intrusa\n");
                return;
            }
        },
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'editado à mão',
    ],
    [
        'nome'     => 'ponteiro da raiz virou fonte paralela',
        'quebra'   => function (string $t) {
            $ps = config()['ponteiros']['arquivos'] ?? [];
            file_put_contents($t . '/' . $ps[0],
                file_get_contents($t . '/' . $ps[0]) . str_repeat("linha extra\n", 60));
        },
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'fonte paralela',
        /*
         * Nem todo teste se aplica a todo projeto. `aplica` é conferido ANTES de rodar.
         *
         * A primeira versão deste arquivo tentava detectar isso pelo retorno da função
         * de quebra — que era null sempre, porque `file_put_contents` no fim do corpo
         * sem `return` devolve null. Resultado: o teste dos ponteiros era pulado em
         * TODO projeto, e o autoteste imprimia sucesso sem nunca ter testado aquilo.
         *
         * Um teste que se pula sozinho em silêncio é pior que um teste ausente.
         */
        'aplica'   => fn (): bool => !empty(config()['ponteiros']['arquivos']),
    ],
    /*
     * ─── cadeia ────────────────────────────────────────────────────────────────
     * Estes testes montam pai e filho do zero, em vez de mexer no que o projeto já tem.
     * Assim funcionam em projeto que ainda não escreveu nenhum PRD — que é o estado de
     * todo projeto no dia da instalação, e justamente quando o autoteste mais precisa
     * funcionar.
     */
    [
        'nome'     => 'deriva_de apontando para documento inexistente',
        'quebra'   => fn (string $t) => escrever($t . '/docs/funcional/_teste.md',
            cabecalhoDeTeste('FDD-TESTE', 'fdd', 'rascunho', ['PRD-QUE-NAO-EXISTE'])),
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'não existe documento com esse id',
        'aplica'   => fn (): bool => !empty(config()['cadeia']['niveis']),
    ],
    [
        'nome'     => 'cadeia invertida — requisito derivando de detalhe',
        'quebra'   => function (string $t) {
            escrever($t . '/docs/arquitetura/_pai.md',
                cabecalhoDeTeste('LLD-TESTE', 'lld', 'rascunho', []));
            escrever($t . '/docs/produto/_filho.md',
                cabecalhoDeTeste('PRD-TESTE', 'prd', 'rascunho', ['LLD-TESTE']));
        },
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'cadeia invertida',
        'aplica'   => fn (): bool => count(config()['cadeia']['niveis'] ?? []) > 1,
    ],
    [
        'nome'     => 'documento vivo pendurado em pai superado',
        'quebra'   => function (string $t) {
            escrever($t . '/docs/produto/_pai.md',
                cabecalhoDeTeste('PRD-TESTE', 'prd', 'superado', []));
            escrever($t . '/docs/funcional/_filho.md',
                cabecalhoDeTeste('FDD-TESTE', 'fdd', 'verificado', ['PRD-TESTE']));
        },
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'superado',
        'aplica'   => fn (): bool => count(config()['cadeia']['niveis'] ?? []) > 1,
    ],
    [
        'nome'     => 'cadeia circular',
        'quebra'   => function (string $t) {
            escrever($t . '/docs/funcional/_a.md',
                cabecalhoDeTeste('FDD-TESTE', 'fdd', 'rascunho', ['ADR-TESTE']));
            escrever($t . '/docs/decisoes/_b.md',
                cabecalhoDeTeste('ADR-TESTE', 'adr', 'rascunho', ['HLD-TESTE']));
            escrever($t . '/docs/arquitetura/_c.md',
                cabecalhoDeTeste('HLD-TESTE', 'hld', 'rascunho', ['FDD-TESTE']));
        },
        'script'   => 'validar-docs.php',
        'esperado' => 1,
        'contem'   => 'cadeia circular',
        'aplica'   => fn (): bool => count(config()['cadeia']['niveis'] ?? []) > 1,
    ],
    [
        'nome'     => 'ciclo não trava o gerador',
        /*
         * Este teste existe porque o laço infinito aconteceu de verdade: o gerador
         * desenhava a árvore recursivamente, encontrou um ciclo, e o processo foi morto
         * pelo sistema depois de consumir a memória da máquina. Sem mensagem, sem pista.
         *
         * Travamento não devolve código de saída útil — então o que se mede aqui é que o
         * processo TERMINA.
         */
        'quebra'   => function (string $t) {
            escrever($t . '/docs/funcional/_a.md',
                cabecalhoDeTeste('FDD-TESTE', 'fdd', 'rascunho', ['ADR-TESTE']));
            escrever($t . '/docs/decisoes/_b.md',
                cabecalhoDeTeste('ADR-TESTE', 'adr', 'rascunho', ['HLD-TESTE']));
            escrever($t . '/docs/arquitetura/_c.md',
                cabecalhoDeTeste('HLD-TESTE', 'hld', 'rascunho', ['FDD-TESTE']));
        },
        'script'   => 'gerar-docs.php',
        'esperado' => 0,
        'contem'   => null,
        'aplica'   => fn (): bool => count(config()['cadeia']['niveis'] ?? []) > 1,
    ],
    [
        'nome'     => 'cadeia bem formada NÃO reprova',
        'quebra'   => function (string $t) {
            escrever($t . '/docs/produto/_pai.md',
                cabecalhoDeTeste('PRD-TESTE', 'prd', 'verificado', []));
            escrever($t . '/docs/funcional/_filho.md',
                cabecalhoDeTeste('FDD-TESTE', 'fdd', 'verificado', ['PRD-TESTE']));
            /*
             * O mapa da cadeia é derivado: acrescentar documento o desatualiza, e o
             * validador reprovaria por isso — não pela cadeia. Regerar antes é o que
             * mantém este teste medindo o que ele diz que mede.
             */
            @exec('cd ' . escapeshellarg($t) . ' && php ' . escapeshellarg(__DIR__ . '/gerar-docs.php') . ' 2>&1');
        },
        'script'   => 'validar-docs.php',
        'esperado' => 0,
        'contem'   => null,
        'aplica'   => fn (): bool => count(config()['cadeia']['niveis'] ?? []) > 1,
    ],
    [
        'nome'     => 'documento novo sem frontmatter aumenta a dívida',
        'quebra'   => fn (string $t) => escrever(
            $t . '/' . trim(config()['acervos']['canonico'], '/') . '/_teste_legado.md',
            "# sem frontmatter\n"),
        'script'   => 'validar-catraca.php',
        'esperado' => 1,
        'contem'   => 'dívida aumentou',
    ],
    [
        'nome'     => 'documento sem frontmatter NÃO é erro de estrutura',
        'quebra'   => fn (string $t) => escrever(
            $t . '/' . trim(config()['acervos']['canonico'], '/') . '/_teste_legado.md',
            "# sem frontmatter\n"),
        'script'   => 'validar-docs.php',
        'esperado' => 0,
        'contem'   => null,
    ],
    [
        'nome'     => 'projeto intacto passa nas três verificações',
        'quebra'   => fn (string $t) => null,
        'script'   => null, // roda as três
        'esperado' => 0,
        'contem'   => null,
    ],
];

// ══════════════════════════════════════════════════════════════ execução
$c = config();
titulo("Autoteste do padrão — {$c['projeto']}");
echo "  cópia de trabalho: {$tmp}\n";
echo "  nada é alterado no seu projeto\n\n";

$passou = $falhou = 0;
$relatos = [];

$pulados = 0;

foreach ($testes as $t) {
    if (isset($t['aplica']) && !($t['aplica'])()) {
        $pulados++;
        $relatos[] = ['·', $t['nome'], 'não se aplica a este projeto (nada configurado para testar)'];
        continue;
    }

    apagar($tmp);
    copiar($origem, $tmp);
    ($t['quebra'])($tmp);

    $scripts = $t['script'] === null
        ? ['validar-docs.php', 'validar-catraca.php', 'validar-indice.php']
        : [$t['script']];

    $rcFinal = 0;
    $saidaFinal = '';
    foreach ($scripts as $s) {
        [$rc, $saida] = rodar($tmp, $s);
        $saidaFinal .= $saida . "\n";
        if ($rc !== 0) {
            $rcFinal = $rc;
        }
    }

    $okCodigo = ($t['esperado'] === 1) ? ($rcFinal !== 0) : ($rcFinal === 0);
    $okTexto  = $t['contem'] === null || str_contains($saidaFinal, $t['contem']);

    if ($okCodigo && $okTexto) {
        $passou++;
        $relatos[] = ['✔', $t['nome'], ''];
    } else {
        $falhou++;
        $motivo = !$okCodigo
            ? ($t['esperado'] === 1 ? 'deveria reprovar e não reprovou' : "reprovou e não deveria (saída {$rcFinal})")
            : "reprovou, mas a mensagem não menciona `{$t['contem']}`";
        $relatos[] = ['✘', $t['nome'], $motivo];
    }
}

apagar($tmp);

foreach ($relatos as [$sinal, $nome, $obs]) {
    printf("  %s  %s%s\n", $sinal, $nome, $obs !== '' ? "\n        → {$obs}" : '');
}

echo "\n" . str_repeat('─', 60) . "\n";
printf("%d passaram · %d falharam%s\n", $passou, $falhou,
    $pulados > 0 ? " · {$pulados} não se aplicam" : '');

if ($falhou > 0) {
    echo "\nUm validador que não reprova o que deveria é pior que nenhum: dá confiança\n";
    echo "falsa. Conserte antes de confiar no build verde.\n";
    exit(1);
}

echo "\nOs validadores pegam o que prometem.\n";
exit(0);

File: memoria_evolutiva/autoteste.py
"""autoteste — os validadores realmente pegam o que prometem?

Verificador que não reprova nada é indistinguível de verificador quebrado — os dois
imprimem "tudo certo". Este comando quebra o projeto DE PROPÓSITO, numa CÓPIA
temporária, uma coisa por vez, e confere que o validador certo reclama. É o teste do
alarme de incêndio: você aperta o botão.

Nada é alterado no seu projeto.
"""

from __future__ import annotations

import json
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

from .lib import config, raiz, titulo

PULAR = {".git", "node_modules", "vendor", "__pycache__", ".code-review-graph"}


def _copiar(de: Path, para: Path) -> None:
    para.mkdir(parents=True, exist_ok=True)
    for item in de.iterdir():
        if item.name in PULAR:
            continue
        destino = para / item.name
        if item.is_dir():
            _copiar(item, destino)
        else:
            try:
                shutil.copy2(item, destino)
            except OSError:
                pass


def _escrever(arquivo: Path, conteudo: str) -> None:
    """Cria o diretório antes — git não versiona pasta vazia, e num CLONE a pasta que
    existe na sua máquina pode simplesmente não estar lá. Quatro testes já falharam no
    CI por isso, passando na máquina de quem escreveu."""
    arquivo.parent.mkdir(parents=True, exist_ok=True)
    arquivo.write_text(conteudo, encoding="utf-8")


def _cabecalho(id_: str, tipo: str, status: str, pais: list[str], projeto: str) -> str:
    fm = (f"---\nid: {id_}\ntipo: {tipo}\nprojeto: {projeto}\n"
          f"titulo: documento de teste do autoteste\nstatus: {status}\n"
          "verificado_em: 2026-01-01\nverificado_commit: 0000000\n")
    if pais:
        fm += "deriva_de:\n" + "".join(f"  - {p}\n" for p in pais)
    return fm + "ancoras:\n  - padrao.json\n---\n\n# documento de teste\n"


def _rodar(tmp: Path, comando: str) -> tuple[int, str]:
    r = subprocess.run([sys.executable, "-m", "memoria_evolutiva", *comando.split()],
                       capture_output=True, text=True, cwd=str(tmp))
    return r.returncode, r.stdout + r.stderr


def main() -> int:
    c = config()
    projeto = c["projeto"]
    origem = Path(raiz())
    ger_dir = c["gerado"].get("diretorio", "docs/gerado").strip("/")
    tem_cadeia = len(c.get("cadeia", {}).get("niveis", {})) > 1
    tem_ponteiros = bool(c.get("ponteiros", {}).get("arquivos"))
    tem_autonomia = bool(c.get("autonomia", {}).get("ativo"))
    tem_grafo = bool(c.get("grafo", {}).get("ativo"))
    tem_fragmentos = isinstance(c.get("rag", {}), dict)
    tem_adaptadores = isinstance(c.get("adaptadores", {}), dict)
    tem_avaliacao = isinstance(c.get("avaliacao", {}), dict)
    tem_ciclo_autonomo = isinstance(c.get("ciclo", {}), dict)
    tem_agendamento = isinstance(c.get("agendamento", {}), dict)

    def quebra_gerado(t: Path) -> None:
        for g in sorted((t / ger_dir).glob("*.md")):
            g.write_text(g.read_text(encoding="utf-8") + "\nlinha intrusa\n", encoding="utf-8")
            return

    def quebra_ponteiro(t: Path) -> None:
        p = t / c["ponteiros"]["arquivos"][0]
        p.write_text(p.read_text(encoding="utf-8") + "linha extra\n" * 60, encoding="utf-8")

    def quebra_ciclo(t: Path) -> None:
        _escrever(t / "docs/funcional/_a.md", _cabecalho("FDD-TESTE", "fdd", "rascunho", ["ADR-TESTE"], projeto))
        _escrever(t / "docs/decisoes/_b.md", _cabecalho("ADR-TESTE", "adr", "rascunho", ["HLD-TESTE"], projeto))
        _escrever(t / "docs/arquitetura/_c.md", _cabecalho("HLD-TESTE", "hld", "rascunho", ["FDD-TESTE"], projeto))

    def quebra_cadeia_ok(t: Path) -> None:
        _escrever(t / "docs/produto/_pai.md", _cabecalho("PRD-TESTE", "prd", "verificado", [], projeto))
        _escrever(t / "docs/funcional/_filho.md", _cabecalho("FDD-TESTE", "fdd", "verificado", ["PRD-TESTE"], projeto))
        # o mapa da cadeia é derivado: regerar antes mantém este teste medindo a cadeia,
        # não a defasagem do derivado
        subprocess.run([sys.executable, "-m", "memoria_evolutiva", "gerar"],
                       capture_output=True, cwd=str(t))

    def quebra_politica_autonoma(t: Path) -> None:
        politica = t / str(c["autonomia"]["politica"])
        politica.write_text(
            politica.read_text(encoding="utf-8").replace(
                "comportamento executado", "preferência arbitrária"
            ),
            encoding="utf-8",
        )

    def quebra_cobertura(t: Path) -> None:
        codigo = t / c.get("gerado", {}).get("raiz", "").strip("/")
        extensoes = {
            "." + str(ext).lower().lstrip(".")
            for ext in c.get("gerado", {}).get("extensoes", [])
        }
        for fonte in sorted(codigo.rglob("*")):
            if fonte.is_file() and fonte.suffix.lower() in extensoes:
                fonte.write_bytes(fonte.read_bytes() + b"\n")
                return

    def quebra_grafo(t: Path) -> None:
        grafo = t / str(c.get("grafo", {}).get("arquivo", "graphify-out/graph.json"))
        grafo.write_text("{corrompido", encoding="utf-8")

    def quebra_fragmentos(t: Path) -> None:
        manifesto = t / str(c.get("rag", {}).get(
            "manifesto", "docs/gerado/manifesto-fragmentos-v2.json"
        ))
        manifesto.write_text(
            manifesto.read_text(encoding="utf-8") + "\n",
            encoding="utf-8",
        )

    def quebra_adaptador(t: Path) -> None:
        caminho = t / ".cursor/rules/memoria-evolutiva.mdc"
        caminho.write_text(
            caminho.read_text(encoding="utf-8").replace("Gateway read-only", "Gateway read-write"),
            encoding="utf-8",
        )

    def quebra_redaction(t: Path) -> None:
        caminho = t / "padrao.json"
        dados = json.loads(caminho.read_text(encoding="utf-8"))
        dados["seguranca_memoria"]["redaction_antes_retain"] = False
        caminho.write_text(
            json.dumps(dados, ensure_ascii=False, indent=4) + "\n", encoding="utf-8"
        )

    def quebra_executor(t: Path) -> None:
        caminho = t / "padrao.json"
        dados = json.loads(caminho.read_text(encoding="utf-8"))
        dados["executor"]["publicacao_automatica"] = True
        caminho.write_text(
            json.dumps(dados, ensure_ascii=False, indent=4) + "\n", encoding="utf-8"
        )

    def quebra_ciclo_autonomo(t: Path) -> None:
        caminho = t / "padrao.json"
        dados = json.loads(caminho.read_text(encoding="utf-8"))
        dados["ciclo"]["publicacao_automatica"] = True
        caminho.write_text(
            json.dumps(dados, ensure_ascii=False, indent=4) + "\n", encoding="utf-8"
        )

    def quebra_agendamento(t: Path) -> None:
        caminho = t / "padrao.json"
        dados = json.loads(caminho.read_text(encoding="utf-8"))
        dados["agendamento"]["banco_negocio"] = "permitido"
        caminho.write_text(
            json.dumps(dados, ensure_ascii=False, indent=4) + "\n", encoding="utf-8"
        )

    def quebra_avaliacao(t: Path) -> None:
        caminho = t / str(c["avaliacao"]["relatorio"])
        caminho.write_text(
            caminho.read_text(encoding="utf-8") + "\n", encoding="utf-8"
        )

    testes: list[dict] = [
        dict(nome="âncora apontando para arquivo inexistente",
             quebra=lambda t: _escrever(t / "docs/arquitetura/_teste.md",
                 _cabecalho("TESTE-ANCORA", "hld", "rascunho", [], projeto).replace(
                     "  - padrao.json", "  - caminho/que/nao/existe.php")),
             comando="validar", esperado=1, contem="âncora quebrada"),
        dict(nome="tipo fora do vocabulário",
             quebra=lambda t: _escrever(t / "docs/arquitetura/_teste.md",
                 _cabecalho("TESTE-VOCAB", "inventado", "rascunho", [], projeto)),
             comando="validar", esperado=1, contem="fora do vocabulário"),
        dict(nome="nome de projeto divergente do padrao.json",
             quebra=lambda t: _escrever(t / "docs/arquitetura/_teste.md",
                 _cabecalho("TESTE-PROJ", "hld", "rascunho", [], "outro-repositorio")),
             comando="validar", esperado=1, contem="não bate com o"),
        dict(nome="id duplicado",
             quebra=lambda t: (
                 _escrever(t / "docs/arquitetura/_a.md", _cabecalho("TESTE-DUP", "hld", "rascunho", [], projeto)),
                 _escrever(t / "docs/arquitetura/_b.md", _cabecalho("TESTE-DUP", "hld", "rascunho", [], projeto))),
             comando="validar", esperado=1, contem="duplicado"),
        dict(nome="arquivo derivado editado à mão",
             quebra=quebra_gerado, comando="validar", esperado=1, contem="editado à mão"),
        dict(nome="política autônoma adulterada",
             quebra=quebra_politica_autonoma, comando="validar", esperado=1,
             contem="política autônoma alterada", aplica=tem_autonomia),
        dict(nome="fonte mudou sem atualizar a cobertura",
             quebra=quebra_cobertura, comando="validar", esperado=1,
             contem="cobertura-codigo.md", aplica=tem_autonomia),
        dict(nome="ponteiro da raiz virou fonte paralela",
             quebra=quebra_ponteiro, comando="validar", esperado=1, contem="fonte paralela",
             aplica=tem_ponteiros),
        dict(nome="deriva_de apontando para documento inexistente",
             quebra=lambda t: _escrever(t / "docs/funcional/_teste.md",
                 _cabecalho("FDD-TESTE", "fdd", "rascunho", ["PRD-QUE-NAO-EXISTE"], projeto)),
             comando="validar", esperado=1, contem="não existe documento com esse id",
             aplica=tem_cadeia),
        dict(nome="cadeia invertida — requisito derivando de detalhe",
             quebra=lambda t: (
                 _escrever(t / "docs/arquitetura/_pai.md", _cabecalho("LLD-TESTE", "lld", "rascunho", [], projeto)),
                 _escrever(t / "docs/produto/_filho.md", _cabecalho("PRD-TESTE", "prd", "rascunho", ["LLD-TESTE"], projeto))),
             comando="validar", esperado=1, contem="cadeia invertida", aplica=tem_cadeia),
        dict(nome="documento vivo pendurado em pai superado",
             quebra=lambda t: (
                 _escrever(t / "docs/produto/_pai.md", _cabecalho("PRD-TESTE", "prd", "superado", [], projeto)),
                 _escrever(t / "docs/funcional/_filho.md", _cabecalho("FDD-TESTE", "fdd", "verificado", ["PRD-TESTE"], projeto))),
             comando="validar", esperado=1, contem="superado", aplica=tem_cadeia),
        dict(nome="cadeia circular",
             quebra=quebra_ciclo, comando="validar", esperado=1, contem="cadeia circular",
             aplica=tem_cadeia),
        # o laço infinito com ciclo aconteceu de verdade: o processo comeu a memória da
        # máquina até ser morto. Aqui se mede que o gerador TERMINA.
        dict(nome="ciclo não trava o gerador",
             quebra=quebra_ciclo, comando="gerar", esperado=0, contem=None, aplica=tem_cadeia),
        dict(nome="cadeia bem formada NÃO reprova",
             quebra=quebra_cadeia_ok, comando="validar", esperado=0, contem=None, aplica=tem_cadeia),
        dict(nome="documento novo sem frontmatter aumenta a dívida",
             quebra=lambda t: _escrever(
                 t / c["acervos"]["canonico"].strip("/") / "_teste_legado.md",
                 "# sem frontmatter\n"),
             comando="catraca", esperado=1, contem="dívida aumentou"),
        dict(nome="documento sem frontmatter NÃO é erro de estrutura",
             quebra=lambda t: _escrever(
                 t / c["acervos"]["canonico"].strip("/") / "_teste_legado.md",
                 "# sem frontmatter\n"),
             comando="validar", esperado=0, contem=None),
        dict(nome="grafo corrompido reprova mesmo com hashes de fonte iguais",
             quebra=quebra_grafo, comando="bancos status --offline", esperado=1,
             contem="marcador do Graphify inválido", aplica=tem_grafo),
        dict(nome="manifesto de fragmentos adulterado reprova",
             quebra=quebra_fragmentos, comando="fragmentos verificar", esperado=1,
             contem="manifesto diverge", aplica=tem_fragmentos),
        dict(nome="adaptador de agente adulterado reprova",
             quebra=quebra_adaptador, comando="adaptadores verificar", esperado=1,
             contem="adaptador", aplica=tem_adaptadores),
        dict(nome="redaction antes do retain não pode ser desativada",
             quebra=quebra_redaction, comando="validar", esperado=1,
             contem="redaction_antes_retain"),
        dict(nome="executor não pode habilitar publicação automática",
             quebra=quebra_executor, comando="validar", esperado=1,
             contem="executor.publicacao_automatica"),
        dict(nome="ciclo dos agentes não pode habilitar publicação automática",
             quebra=quebra_ciclo_autonomo, comando="validar", esperado=1,
             contem="ciclo.publicacao_automatica", aplica=tem_ciclo_autonomo),
        dict(nome="agendador não pode acessar o banco de negócio",
             quebra=quebra_agendamento, comando="validar", esperado=1,
             contem="agendamento.banco_negocio", aplica=tem_agendamento),
        dict(nome="relatório de avaliação RAG adulterado reprova",
             quebra=quebra_avaliacao, comando="avaliar verificar", esperado=1,
             contem="relatório RAG", aplica=tem_avaliacao),
        dict(nome="projeto intacto passa nas verificações locais",
             quebra=lambda t: None, comando=None, esperado=0, contem=None),
    ]

    titulo(f"Autoteste do padrão — {projeto}")
    print("  cópia de trabalho: temporária; nada é alterado no seu projeto\n")

    passou = falhou = pulados = 0
    relatos: list[tuple[str, str, str]] = []

    # A cópia deliberadamente não leva `.git` nem dependências. Refaça uma vez os
    # derivados dependentes da raiz/commit e os adaptadores com caminho absoluto; sem
    # essa preparação, um projeto Git com `vendor/` falha intacto e cria falso positivo.
    with tempfile.TemporaryDirectory(prefix="autoteste-base-") as base_s:
        base = Path(base_s) / "projeto"
        _copiar(origem, base)
        for cmd in ("gerar", "adaptadores gerar"):
            rc_base, saida_base = _rodar(base, cmd)
            if rc_base != 0:
                print(f"\nNão foi possível preparar a cópia-base com `{cmd}`:\n{saida_base}")
                return 1

        for t in testes:
            if not t.get("aplica", True):
                pulados += 1
                relatos.append(("·", t["nome"], "não se aplica a este projeto"))
                continue

            with tempfile.TemporaryDirectory(prefix="autoteste-memoria-") as tmp_s:
                tmp = Path(tmp_s) / "projeto"
                _copiar(base, tmp)
                t["quebra"](tmp)

                comandos = ([t["comando"]] if t["comando"]
                            else ["validar", "catraca", "bancos status --offline",
                                  "avaliar verificar"])
                rc_final, saida_final = 0, ""
                for cmd in comandos:
                    rc, saida = _rodar(tmp, cmd)
                    saida_final += saida
                    if rc != 0:
                        rc_final = rc

            ok_codigo = (rc_final != 0) if t["esperado"] == 1 else (rc_final == 0)
            ok_texto = t["contem"] is None or t["contem"] in saida_final

            if ok_codigo and ok_texto:
                passou += 1
                relatos.append(("✔", t["nome"], ""))
            else:
                falhou += 1
                motivo = ("deveria reprovar e não reprovou" if t["esperado"] == 1 and not ok_codigo
                          else f"reprovou e não deveria (saída {rc_final})" if not ok_codigo
                          else f"reprovou, mas a mensagem não menciona `{t['contem']}`")
                relatos.append(("✘", t["nome"], motivo))

    for sinal, nome, obs in relatos:
        print(f"  {sinal}  {nome}" + (f"\n        → {obs}" if obs else ""))

    print("\n" + "─" * 60)
    extra = f" · {pulados} não se aplicam" if pulados else ""
    print(f"{passou} passaram · {falhou} falharam{extra}")

    if falhou:
        print("\nUm validador que não reprova o que deveria é pior que nenhum: dá confiança")
        print("falsa. Conserte antes de confiar no build verde.")
        return 1
    print("\nOs validadores pegam o que prometem.")
    return 0
